### PR TITLE
Metrics for SDF

### PR DESCRIPTION
**Pull request by vodovozovaliza**
Implemented metrics for SDF from NN:
- IoU
- Hausdorf
- Chamfer

### CODE REVIEW

File: src/nlotrajectories/core/metrics.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    target_mask = sdf_target < threshold
+    prediction_mask = sdf_pred < threshold
+
+    intersection = np.logical_and(target_mask, prediction_mask).sum()
+    union = np.logical_or(target_mask, prediction_mask).sum()
+
+    if union == 0:
+        return 1.0 if intersection == 0 else 0.0
+
+    iou = intersection / union
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/metrics.py:41]:
```suggestion
    return intersection / union
```

File: src/nlotrajectories/core/metrics.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    return iou
+
+
+def chamfer(sdf_target: np.array, sdf_pred: np.array, X: np.array, Y: np.array, eps: float = 1e-2):
+    # compute the chamfer distance between sdf_target and sdf_pred
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/metrics.py:46]: No need for second comment

File: src/nlotrajectories/core/metrics.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    union = np.logical_or(target_mask, prediction_mask).sum()
+
+    if union == 0:
+        return 1.0 if intersection == 0 else 0.0
+
+    iou = intersection / union
+    return iou
+
+
+def chamfer(sdf_target: np.array, sdf_pred: np.array, X: np.array, Y: np.array, eps: float = 1e-2):
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/metrics.py:45]:
```suggestion
def chamfer(sdf_target: np.ndarray, sdf_pred: np.ndarray, X: np.ndarray, Y: np.ndarray, eps: float = 1e-2):
```

File: src/nlotrajectories/core/metrics.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    if pred_points.shape[0] == 0 or target_points.shape[0] == 0:
+        return float("inf")  # No surface points to compare
+
+    # Compute the chamfer distance between pred_points and target_points
+    dists = np.linalg.norm(pred_points[:, None] - target_points[None, :], axis=-1)
+    chamfer_distance = (np.mean(np.min(dists, axis=1)) + np.mean(np.min(dists, axis=0))) / 2
+    return chamfer_distance
+
+
+def surface_loss(sdf_target: np.array, sdf_pred: np.array, X: np.array, Y: np.array, eps: float = 1e-2):
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/metrics.py:77]:
```suggestion
def surface_loss(sdf_target: np.ndarray, sdf_pred: np.ndarray, X: np.ndarray, Y: np.ndarray, eps: float = 1e-2):
```

File: src/nlotrajectories/core/sdf/l4casadi.py [hunk 5]
```diff
@@ -101,7 +103,18 @@ def train(
...
                 xb, yb = xb.to(self.device), yb.to(self.device)
                 pred = self.model(xb)
-                loss = loss_fn(pred, yb)
+
+                # Compute the MSE loss
+                mse_loss = loss_fn(pred, yb)
+
+                # Compute the surface loss
+                if surface_loss_weight > 0:
+                    surface_mask = torch.abs(yb) < surface_loss_eps
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/sdf/l4casadi.py:112]: Why not use surface_loss that you already implemented?

File: src/nlotrajectories/core/visualizer.py [hunk 6]
```diff
@@ -46,3 +47,112 @@ def plot_levels(func, x_range=(-1, 2), y_range=(-1, 2), n_samples=500, title="sd
     plt.axis("equal")
     plt.savefig(path / f"{title}_levels.png", bbox_inches="tight")
     plt.close()
+
+
+def plot_control(U_opt, dt: float, title="Control Inputs"):
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/visualizer.py:52]:
```suggestion
def plot_control(U_opt, dt: float, title: str = "Control Inputs"):
```

File: src/nlotrajectories/core/sdf/l4casadi.py [hunk 7]
```diff
@@ -101,7 +104,25 @@ def train(
...
                 xb, yb = xb.to(self.device), yb.to(self.device)
                 pred = self.model(xb)
-                loss = loss_fn(pred, yb)
+
+                # Compute the MSE loss
+                mse_loss = loss_fn(pred, yb)
+                # Compute the surface loss
+                if surface_loss_weight > 0:
+                    surface_loss_value = surface_loss(
+                        yb.detach().cpu().numpy(),
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/sdf/l4casadi.py:113]: With detach gradient can not pass through this loss
```suggestion
                        yb
```

File: src/nlotrajectories/core/sdf/l4casadi.py [hunk 8]
```diff
@@ -101,7 +104,25 @@ def train(
...
+                        pred.detach().cpu().numpy(),
+                        xb[:, 0].detach().cpu().numpy(),
+                        xb[:, 1].detach().cpu().numpy(),
+                        eps=surface_loss_eps,
+                    )
+                    surface_loss_value = torch.tensor(surface_loss_value, dtype=torch.float32, device=self.device)
+                    # if surface_loss_value is NaN because not common surface points, set it to 0.0
+                    if torch.isnan(surface_loss_value):
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/sdf/l4casadi.py:121]:
```suggestion
                    if not torch.isnan(surface_loss_value):
                        loss += surface_loss_weight * surface_loss_value
```

File: src/nlotrajectories/core/sdf/l4casadi.py [hunk 9]
```diff
@@ -128,7 +149,22 @@ def train(
                 for xb, yb in val_loader:
                     xb, yb = xb.to(self.device), yb.to(self.device)
                     pred = self.model(xb)
-                    val_loss += loss_fn(pred, yb).item() * xb.size(0)
+                    if surface_loss_weight > 0:
+                        surface_loss_value = surface_loss(
+                            yb.detach().cpu().numpy(),
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/sdf/l4casadi.py:154]: Here you can also delete `.detach().cpu().numpy()`

File: src/nlotrajectories/core/metrics.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    if pred_points.shape[0] == 0 or target_points.shape[0] == 0:
+        return None  # No surface points to compare
+
+    # Compute the chamfer distance between pred_points and target_points
+    dists = np.linalg.norm(pred_points[:, None] - target_points[None, :], axis=-1)
+    chamfer_distance = (np.mean(np.min(dists, axis=1)) + np.mean(np.min(dists, axis=0))) / 2
+    return chamfer_distance
+
+
+def surface_loss(sdf_target: np.ndarray, sdf_pred: np.ndarray, eps: float = 1e-2) -> float | None:
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/metrics.py:75]: sdf_target and sdf_pred can also be tensor

File: src/nlotrajectories/core/metrics.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    """
+    Compute the surface loss of the approximated sdf
+    Args:
+        sdf_target (np.ndarray): Ground truth SDF grid.
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/metrics.py:79]: Same here

File: src/nlotrajectories/core/metrics.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        sdf_pred (np.ndarray): Predicted SDF grid.
+        X (np.ndarray): X coordinates of the grid.
```
Comment (IavorskaiaMariia) [inline src/nlotrajectories/core/metrics.py:81]: Didn't delete description of missing arguments
